### PR TITLE
[DataFrame] Apply() for Lists and Dicts

### PR DESCRIPTION
**Pull request by osalpekar**

### CODE REVIEW

File: python/ray/dataframe/dataframe.py [hunk 1]
```diff
@@ -996,24 +987,56 @@ def apply(self, func, axis=0, broadcast=False, raw=False, reduce=None,
...
+            if has_list:
+                # if input dict has a list, the function to apply must wrap
+                # single functions in lists as well to get the desired output
+                # format
+                result = [_deploy_func.remote(
+                    lambda df: df.iloc[:, tup[0][1]].apply(
+                        func[tup[1]] if is_list_like(func[tup[1]])
+                        else [func[tup[1]]]),
+                    self._col_partitions[tup[0][0]])
+                    for tup in part_ind_tuples]
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: You can auto-unbox this by replacing `tup` with `(p, iwp), key` or something like that.

File: python/ray/dataframe/dataframe.py [hunk 2]
```diff
@@ -996,24 +987,56 @@ def apply(self, func, axis=0, broadcast=False, raw=False, reduce=None,
...
             if axis == 1:
                 kwds['axis'] = axis
             return getattr(self, func)(*args, **kwds)
+        elif isinstance(func, dict):
+            if axis == 1:
+                raise TypeError(
+                    "(\"'dict' object is not callable\", "
+                    "'occurred at index {0}'".format(self.index[0]))
+            has_list = list in map(type, func.values())
+            part_ind_tuples = [(self._col_metadata[key], key) for key in func]
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: Do we define behavior with non-unique columns? That will cause `self._col_metadata[key]` to result in a DataFrame and not a Series.

File: python/ray/dataframe/dataframe.py [hunk 3]
```diff
@@ -996,24 +987,56 @@ def apply(self, func, axis=0, broadcast=False, raw=False, reduce=None,
...
+                        func[tup[1]] if is_list_like(func[tup[1]])
+                        else [func[tup[1]]]),
+                    self._col_partitions[tup[0][0]])
+                    for tup in part_ind_tuples]
+                return pd.concat(ray.get(result), axis=1)
+            else:
+                result = [_deploy_func.remote(
+                    lambda df: df.iloc[:, tup[0][1]].apply(func[tup[1]]),
+                    self._col_partitions[tup[0][0]])
+                    for tup in part_ind_tuples]
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: Same as above.

File: python/ray/dataframe/index_metadata.py [hunk 4]
```diff
@@ -52,7 +52,7 @@ def _set_index(self, new_index):
     index = property(_get_index, _set_index)
 
     def coords_of(self, key):
-        raise NotImplementedError()
+        self._coord_df.loc[key]
```
Comment (Veryku) [inline python/ray/dataframe/index_metadata.py]: You shouldn't need to change this in `_IndexMetadataBase`, it should be implemented in `_IndexMetadata`. Either way, `_IndexMetadataBase` is about to be deprecated by #1965

File: python/ray/dataframe/dataframe.py [hunk 5]
```diff
@@ -996,24 +987,57 @@ def apply(self, func, axis=0, broadcast=False, raw=False, reduce=None,
...
+                    "'occurred at index {0}'".format(self.index[0]))
+            # TODO: some checking on functions that return Series or Dataframe
+            new_cols = _map_partitions(lambda df: df.apply(func),
+                                       self._col_partitions)
+
+            # resolve function names for the DataFrame index
+            new_index = [f_name if isinstance(f_name, compat.string_types)
+                         else f_name.__name__ for f_name in func]
+            return DataFrame(col_partitions=new_cols,
+                             columns=self.columns,
```
Comment (Veryku) [inline python/ray/dataframe/dataframe.py]: Add `col_metadata=self._col_metadata` after #1965

**Review by p-yang — APPROVED** [review]
LGTM but let's wait for #1965 to get merged in to rebase and change the constructor.

File: python/ray/dataframe/dataframe.py (hunk 5 above)
Comment (devin-petersohn) [reply python/ray/dataframe/dataframe.py]: ping @osalpekar. Rebase and make this change, then looks great!

Comment (osalpekar) [reply python/ray/dataframe/dataframe.py]: rebased and made change in commit below!

**Review by devin-petersohn — APPROVED** [review]

**Comment by devin-petersohn** [comment]
Jenkins, retest this please.

**Comment by devin-petersohn** [comment]
Jenkins, retest this please.

**Comment by devin-petersohn** [comment]
Merged, thanks @osalpekar

**Pull request merged**